docs: Add more background info for parallelism page (#188)
bentoml/llm-inference-in-production: docs/inference-optimization/data-tensor-pipeline-expert-hybrid-parallelism.md

## Data parallelism

> Data parallelism (DP) runs multiple replicas of the same model in parallel. Each

Reviewer: I actually like the way the previous version states the purpose. I suggest you bring a version of that back into the first sentence. I think, for a page like this that is a miniature glossary, we should treat the first sentence like the first sentence of a page—it should work well as a description snippet. Maybe something like:

```suggestion
Data parallelism (DP) increases total throughput by distributing individual
requests or microbatches across multiple replicas of the same model. Each
```

request, because that request is still processed by a single replica.

Data parallelism also provides a failure boundary. When a health check marks one
replica unavailable, the router can stop sending new requests to it and use the
remaining replicas. This improves availability, though the healthy replicas need
spare capacity to absorb the traffic. Any in-flight request and local KV cache
on the failed replica may also be lost.

A key consideration here is how to route requests. For LLM inference, requests
can vary widely in prompt length, output length, and KV cache usage. GPU
capacity is also expensive, so poor load distribution can leave some replicas
overloaded while others sit underused. Production routers therefore often
consider multiple signals when choosing a replica. For more information, see
[inference routing](/inference-optimization/inference-routing/).

The main cost of data parallelism is memory duplication. Every replica needs a
copy of the model weights and maintains a separate KV cache. If one copy of the
model can't fit on a worker, data parallelism alone doesn't solve the problem.
Combine it with tensor or pipeline parallelism.

## Tensor parallelism

> Today, many LLMs can't fit on a single GPU, even after quantization, especially

Reviewer: Same here: I think the first sentence should provide a quick definition. As is, for somebody with a reasonably good understanding of LLMs, the answer to "what is tensor parallelism" is buried. I think it’s better if we provide a quick definition that’s effective for most of our readers, and then elaborate for folks who know less—instead of the reverse which assumes the reader knows very little and potentially over-explains the background information before telling the reader what TP is.

For example:

```suggestion
Tensor parallelism (TP) distributes the weights of a model across
multiple GPUs in order to deploy large models that can’t fit into
the memory of a single GPU.

Today, many LLMs can't fit on a single GPU, even after quantization, especially
```

divides the tensors within each model layer across multiple GPUs. During
operations such as matrix multiplication, each GPU holds a shard of the weight
tensor and computes part of the result. The GPUs then exchange or combine
intermediate results using collective communication operations such as
all-reduce or all-gather, depending on how the tensors are partitioned.

<Diagram name="tp-inference" alt="Tensor parallelism: model layers split across GPUs" />

By distributing weights, cache, and computation, tensor parallelism makes models
that exceed the capacity of a single GPU practical to serve. It can also lower
the latency of an individual request when the gains from parallel computation
outweigh the added communication cost.

That communication is the central tradeoff. GPUs in tensor parallelism must
exchange intermediate results repeatedly as a request flows through the
transformer layers. Interconnect bandwidth and latency therefore weigh heavily
on performance.

Tensor parallelism works best among GPUs linked by high-bandwidth
interconnects within a single node, such as NVLink. Stretching a high tensor

> parallelism degree across nodes can turn network communication into the dominant

Reviewer: I don’t know what "degree" means in this sentence, so this sentence is confusing. Also, "high tensor parallelism degree" is kind of a mouthful of a concept. Is there another way you can say this? For example, does it work if you instead say, "Parallelizing a very large tensor across nodes can turn..."? Please reconsider the next sentence as well to remove "degree" unless you can gracefully define that (but I don’t think you need to and can probably just avoid using it).

Author: I went with "splitting tensors across GPUs that sit in separate nodes" here because the bottleneck isn't really about tensor size, it's about where the GPUs are. A modest tensor split across two nodes still pays the network cost, while a huge one split within a node over NVLink may be fine.

bottleneck. The degree is also constrained; it generally has to divide the
number of attention heads evenly, so you can't scale to arbitrary GPU counts.

> Before sharding a model, check whether the weights, KV cache, and other

Reviewer: Just to be clear, I think you should say, "Before sharding a model to perform tensor parallelism, check..."

Author: Make sense

inference memory requirements fit comfortably on a single GPU, especially after
[quantization](/model-preparation/llm-quantization/). Single-GPU serving avoids
cross-device synchronization entirely, and adding GPUs through tensor
parallelism rarely yields a linear speedup. Use tensor parallelism when the
memory footprint or latency targets justify the communication overhead.

## Pipeline parallelism

Pipeline parallelism (PP) divides the model’s layers into sequential chunks,
each assigned to a separate device. Data flows through these chunks like an
assembly line, with the output of one device becoming the input for the next.
For instance, in a four-way pipeline, each device processes a quarter of the
model’s layers.

<Diagram name="pp-diagram" alt="Pipeline parallelism: consecutive layer ranges on each GPU" />

> Unlike tensor parallelism, pipeline parallelism does not require devices to

Reviewer: Minor: our house style prefers contractions, because they create a more casual, conversational voice.

```suggestion
Unlike tensor parallelism, pipeline parallelism doesn’t require devices to
```

combine partial results within every layer. A stage sends activations to the

> next stage only after completing the assigned layer range. This lower

Reviewer: I’m not sure what "layer range" means. Can you instead just say "the assigned layers"?

Author: Sorry for the confusion..."layer range" means the consecutive block of layers assigned to one device, e.g. layers 1–20 on GPU 0, 21–40 on GPU 1. I used the term "stage" here without connecting to it. Let me tie the vocabulary together so one term is defined up front and reused.

communication frequency can make pipeline parallelism a better fit for slower

> links, including links between nodes.

Reviewer: "slower links" is ambiguous here. Maybe it’d be more clear if you use the same terminology from the TP section. For example, "for low bandwidth connections, including slow interconnects between nodes."
